book/src/developers/designs/application_passwords.md

```diff
@@ -0,0 +1,188 @@
+# Rationale
+
+Kanidm exists to provide an authentication source for external applications.
+These applications need to have standardised ways to integrate with Kanidm to
+allow that application to interact and trust Kanidm's authentication results.
+
+For web based applications we offer Oauth2/OIDC. For Linux machines we offer a
+Kanidm specific HTTPS channel for identifying users. We also offer an LDAPS
+gateway for applications that don't support other protocols but we want users to
+use their "UNIX password".
+
+However this has the issue that due to how limited Linux authentication is, UNIX
+passwords are always single factor. We also don't want the same UNIX credentials
+that are used, e.g. for sudo on machines, to be leaked and used for applications
+like email.
+
+To improve this we need a way to offer authentication services to applications
+that are unable to support anything modern.
+
+Since LDAP is the "lingua franca" of authentication and almost universally
+implemented as an authentication for all applications, we can use this to
+provide *application specific* password based authentication.
+
+# User experience
+
+The administrator configures two applications on their Kanidm instance. One is
+"mail" for a generic SMTP+IMAP service. The other is HTTP basic auth to a legacy
+web server. Once applications are added, the administrator adds the allowed
+users to the applications as if they were regular groups.
+
+The mail services and web services are configured to point to Kanidm's LDAP
+gateway with a customized search base DN.
+
+The user can login to the webui or command line and list what linked
+applications exist on their account that require application passwords.
+
+The user can then request a new "application password" for the mail server for
+their laptop. They can copy-paste the generated password to their mail client
+which uses this password on their behalf. Similar he can request a new
+application password to access the web server basic auth.
+
+As a user can request several application passwords for each application, the
+username will have a label appended like "username@laptop" or "username@phone".
```

Comment:
The username@app format's a problem given we already use SPN's which are `user@domain`, maybe `username+app` ?

Reply:
There actually isn't a need for this at all. 

Because the imap server is configured to ldap `app=imap,dc=idm,dc=example,dc=com`, and http to `app=http,dc=idm,dc=example,dc=com` then we can use username/spn as usual. The distinction/routing of the application password to use *isn't part of the username*. It's the applications ldap basedn that routes to the specific application that we are interacting with at this point in time.


As a more concrete example:

```
ldapwhoami -D name=william,app=imap,dc=idm,dc=example,dc=com -w "imap-application-password-here"

ldapwhoami -D name=william,app=http,dc=idm,dc=example,dc=com -w "http-application-password-here"
```

That's why both imap/http would have the search basedns move to app=imap,dc=idm,dc=example,dc=com or app=http,dc=idm,dc=example,dc=com

Reply:
It is not the app what is appended but the label. As a user can have multiple passwords for the same application it helps to identify which one a client application is using. For example, a user configures his mail client on his phone and his laptop and several months later he wants to delete the phone's one, but he was not careful enough to properly label the passwords at creation time. If the label is appended to the username, he could open the mail application settings on the phone and check the username to identify the app password to remove on the server.

This is just an idea I took from Roundcube because it is useful to myself, but if we agree it is not convenient no problem on dropping it.

Reply:
When it would be confusing is our eventual trusts setup because then you'd have username@domain@label. So I think it gets pretty chaotic there. 

I'm looking at other systems like fastmail, and they have the labels as something for the user in the webui to know what each application password was used for, rather than it being a part of the username. I think username@label being tied to an app pw would create some confusion to what users have to enter.

So lets drop the @label here I think.